--- a/BUILD	2020-03-05 00:22:13.000000000 -0500
+++ b/BUILD	2020-03-05 00:22:25.000000000 -0500
@@ -60,7 +60,7 @@
 posix_copts = [
     # Assembler option --noexecstack adds .note.GNU-stack to each object to
     # ensure that binaries can be built with non-executable stack.
-    "-Wa,--noexecstack",
+    # "-Wa,--noexecstack",
 
     # This is needed on Linux systems (at least) to get rwlock in pthread.
     "-D_XOPEN_SOURCE=700",
